source/common/upstream/subset_lb.cc

```diff
@@ -340,31 +356,63 @@ bool SubsetLoadBalancer::hostMatches(const SubsetMetadata& kvs, const Host& host
 
 // Iterates over subset_keys looking up values from the given host's metadata. Each key-value pair
 // is appended to kvs. Returns a non-empty value if the host has a value for each key.
-SubsetLoadBalancer::SubsetMetadata
+std::vector<SubsetLoadBalancer::SubsetMetadata>
 SubsetLoadBalancer::extractSubsetMetadata(const std::set<std::string>& subset_keys,
                                           const Host& host) {
-  SubsetMetadata kvs;
+  std::vector<SubsetMetadata> all_kvs;
 
   const envoy::api::v2::core::Metadata& metadata = *host.metadata();
   const auto& filter_it = metadata.filter_metadata().find(Config::MetadataFilters::get().ENVOY_LB);
   if (filter_it == metadata.filter_metadata().end()) {
-    return kvs;
+    return all_kvs;
   }
 
   const auto& fields = filter_it->second.fields();
   for (const auto key : subset_keys) {
     const auto it = fields.find(key);
     if (it == fields.end()) {
+      all_kvs.clear();
       break;
     }
-    kvs.emplace_back(std::pair<std::string, ProtobufWkt::Value>(key, it->second));
-  }
 
-  if (kvs.size() != subset_keys.size()) {
-    kvs.clear();
+    if (list_as_any_ && it->second.kind_case() == ProtobufWkt::Value::kListValue) {
+      // If the list of kvs is empty, we initialize one kvs for each value in the list.
+      // Otherwise, we branch the list of kvs by generating one new kvs per old kvs per
+      // new value.
+      //
+      // For example, two kvs (<a=b>, <c=d>) joined with the kv foo=[bar,baz] results in four kvs:
+      //   <a=b,foo=bar>
+      //   <a=b,foo=baz>
+      //   <c=d,foo=bar>
+      //   <c=d,foo=baz>
```

Comment:
Minor: I think this only happens when list-valued metadata are encountered multiple times, so maybe the example should for two kvs (<a=1>, <a=2>)?

Reply:
Yeah you're right, I'll update the comment. In my mind I was conflating a single kvs with two entries from two non-list entries with two kvs entries from a single list entry.